src/OpenTelemetry.AutoInstrumentation/Instrumentations/MongoDb/MongoClientIntegration.cs

```diff
@@ -0,0 +1,151 @@
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { "MongoDB.Driver.MongoClientSettings" },
+        MinimumVersion = "2.3.0",
+        MaximumVersion = "2.65535.65535",
+        IntegrationName = "MongoDb")]
+    public class MongoClientIntegration
+    {
+#if NETCOREAPP3_1_OR_GREATER
+        private static Delegate _setActivityListener;
+#endif
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TMongoClientSettings">Type of the settings</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="settings">The source of the original GraphQL query</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TMongoClientSettings>(TTarget instance, TMongoClientSettings settings)
+        {
+#if NETFRAMEWORK
+            // Additional deps doesn't support .NET FX
+            // TODO: Find another way how to ship "MongoDB.Driver.Core.Extensions.DiagnosticSources"
+            return CallTargetState.GetDefault();
+#endif
+
+#if NETCOREAPP3_1_OR_GREATER
+            Assembly driverExtensions;
+            Type driverAcitivitySubscriber = Type.GetType("MongoDB.Driver.Core.Extensions.DiagnosticSources.DiagnosticsActivityEventSubscriber, MongoDB.Driver.Core.Extensions.DiagnosticSources");
+            if (driverAcitivitySubscriber == null)
+            {
+                string basePath = Environment.GetEnvironmentVariable("DOTNET_SHARED_STORE");
+                string processArch = Environment.Is64BitProcess ? "x64" : "x32";
+                string dllPath = Path.Combine("netcoreapp3.1", "mongodb.driver.core.extensions.diagnosticsources", "1.1.0", "lib", "netstandard2.0", "MongoDB.Driver.Core.Extensions.DiagnosticSources.dll");
+                string assemblyPath = Path.Combine(basePath, processArch, dllPath);
+
+                if (!File.Exists(assemblyPath))
+                {
+                    return CallTargetState.GetDefault();
+                }
+
+                driverExtensions = Assembly.LoadFile(assemblyPath);
+            }
+            else
+            {
+                driverExtensions = Assembly.GetAssembly(driverAcitivitySubscriber);
+            }
+
+            Type optionsType = driverExtensions.GetType("MongoDB.Driver.Core.Extensions.DiagnosticSources.InstrumentationOptions");
+
+            if (optionsType != null)
+            {
+                var setListenerAction = _setActivityListener ??
+                    (_setActivityListener = GetClusterConfiguratorExpression(driverExtensions, optionsType).Compile());
+
+                settings
+                   .GetType()
+                   .GetProperty("ClusterConfigurator", BindingFlags.Public | BindingFlags.Instance)
+                   .SetValue(settings, setListenerAction);
```

Comment:
Should we combine the delegates so we don't overwrite a developer's existing delegate?
```csharp
                var property = settings.GetType().GetProperty("ClusterConfigurator", BindingFlags.Public | BindingFlags.Instance);
                var existingDelegate = property.GetValue(settings) as Delegate;
                property.SetValue(settings, Delegate.Combine(existingDelegate, setListenerAction));
```

Reply:
Thanks, wanted to keep it later but as you already provided the solution, I'm going to include it.